### DISCUSSION TITLE
Mirth Javascript replacement not working

### DISCUSSION THREAD

**Discussion by mayowan16** [nextgenhealthcare/connect#5719] [Q&A]
We are on Mirth Connect 4.0.1.  

I am attempting to simply copy PV1.7 to PV1.8 when PV1.8 is empty and the message type is an A04 or A08.  I have the exact same replacement code elsewhere is my channel and it working perfectly.  However, when I try to carry out this specific change only from PV1.7, the output is in XML format.  I copied the second row of the replacement from a different transformer to test it in this one, and the output from PV1.9 came out correctly.  You can see that there is XML where the PV1.7 was copied, but not where the PV1.9 was copied.

"<PV1.7.1>1770783607</PV1.7.1><PV1.7.1>13029</PV1.7.1>^BOONE".  

Here is the transformer script.
//Copy Rendering to Referring if PV1.8 is blank.

if (msg['MSH']['MSH.9']['MSH.9.2'].toString() == "A04" && (!msg['PV1']['PV1.8']['PV1.8.1'].toString()))
			
	{
		msg['PV1']['PV1.8']['PV1.8.1'] = msg['PV1']['PV1.7']['PV1.7.1'].toString();
		msg['PV1']['PV1.8']['PV1.8.2'] = msg['PV1']['PV1.9']['PV1.9.2'].toString();
	
	
	}
	else 
		{
			;
		}
I also tried to create a mapper for the entire PV1.7 segment and replace it.  The same XML issue occured.  Here is the script for that attempted step.

msg['PV1']['PV1.8'] = $('AttendingProviderPV1.7WholeSeg'); 

I read another forum that said PV1.7 might be switching to XML in a prior step.  I have checked the destinations on the source channel, and all of my outputs are message.encoundedData, not transformed.  I'm not sure what to look at next.

Thanks in advance!

**Comment by jonbartels** [comment]
It looks like PV1.7 is a repeating segment. Can you post a sample PV1 segment from the inbound message please.

**Comment by mayowan16** [comment]
Wow.  Good call.  I did not realize the sender was doing this.

1770783607^BOONE^CHRISTOPHER^^^^^^NPI^^^^NPI~13029^PRP^POSM^BELL ORTHO^^^^^^^^^

I only need to use the first repetition.

**Comment by mayowan16** [comment]
I tried creating a source transformer to capture the first PV1.7[0].  This got rid of the repetition data, but it is still coming out as XML.  I'm not sure what to do to make it a string.
![image](https://user-images.githubusercontent.com/125709147/225136689-a5ff3099-b1bb-4b2c-b729-fb91cdd8a0db.png)
This is the most recent output.  
<PV1.7><PV1.7.1>1770783607</PV1.7.1><PV1.7.2>BOONE</PV1.7.2><PV1.7.3>CHRISTOPHER</PV1.7.3><PV1.7.4/><PV1.7.5/><PV1.7.6/><PV1.7.7/><PV1.7.8/><PV1.7.9>NPI</PV1.7.9><PV1.7.10/><PV1.7.11/><PV1.7.12/><PV1.7.13>NPI</PV1.7.13></PV1.7>
This is my edited Javascript.
if (msg['MSH']['MSH.9']['MSH.9.2'].toString() == "A04" && (!msg['PV1']['PV1.8']['PV1.8.1'].toString()))
		
	{
			msg['PV1']['PV1.8'] = $('patientVisit_attendingDoctor').toString();
	}

**Reply by jonbartels** [reply]
`msg['PV1']['PV1.7'][0]['PV1.7.1'].toString();` PV1 segment, first repetition of PV1.7, then the first field of the first repetition of PV1.7.1

**Reply by mjtrac** [reply]
To generalize what Jon is pointing out, when you use .toString() on anything that has child nodes, you will get the XML tags along with any values within.  When you use .toString() on something that has no child nodes, you will get the value itself.  There is also a .toXMLString() method which will always return the tags surrounding the value.  msg['PV1']['PV1.7'] is a list of the repeats of the PV1.7; msg['PV1']['PV1.7'][0] is the first repetition, but it still has components underneath it in the tree.

You can explore this, if you wish, by running the Rhino JS interpreter included in Mirth Connect.  From a terminal, go to the root of your Mirth installation, then down to server-lib.  Type java -jar rhino...jar (I forget the exact name of the rhino jar, but it begins "rhino.")  That will put you into the JS interpreter, and you can load XML into it like this: var msg = [a]the a node[b]the b node[c]the c node[/c][/b][/a]; -- but use angle brackets rather than square brackets I've used to keep the brackets visible in this editor.  msg points to the a node, msg['b'] points to the b node, msg['c'] to c.  You can then explore what .toString() and .toXMLString() return from each.
